readthedocs/api/v2/serializers.py

```diff
@@ -98,7 +98,30 @@ class Meta(ProjectSerializer.Meta):
 
 
 class VersionSerializer(serializers.ModelSerializer):
-    project = ProjectSerializer()
+
+    """
+    Version serializer.
+
+    Instead of using directly a ProjectSerializer for the project,
+    we user a SerializerMethodField to have more control over the
+    serialization of the project, this allows us to optimize the
+    serialization of the same project for each version.
+
+    We usually filter all versions that belong to one project,
+    so instead of serializing the same project over and over again,
+    we cache the serialized project and reuse it for each version.
+
+    Why not just rely on select_related('project')?
+    Since the project is the same for all versions most of the time,
+    we would be serializing the same project over and over again,
+    and since the ProjectSerializer includes a call to get_docs_url,
+    we would be calling get_docs_url for the same project over and over again,
+    generating more queries.
```

Comment:
I think it makes sense to give it a try to the cached version of `get_docs_url` first. This will also help in other scenarios too and the code/flow will be simpler.

Reply:
So, I gave it a try but:

We need to cache the resolve function, since if we cache get_docs_url, that's per-project instance, resulting in not caching all other projects. But if we do this lot of tests break, we can fix this by purging the cache each time after each test, but the next other problem makes this effort in vain.

The project serializer also access other fields from the project: `users` and `features` (they cause extra queries per version), users is easy to add to `prefetch_related`, but `features` is a property with a custom queryset, so it can't be added

https://github.com/readthedocs/readthedocs.org/blob/7f643d9298a54fe5e7e1693059bd748b8109898e/readthedocs/projects/models.py#L1292-L1294


Maybe, maybe we would be able to use a custom prefetch object, but that feels more complicated than just caching the serialized project, plus, caching the serializer we gain some speed from not serializing the same project more than once.

Reply:
OK. Thanks for the explanation on this. Would you mind adding a comment in the code about this that also points to this conversation. This will give us context in the future when I ask the same question 😄